easypermissions/src/main/java/pub/devrel/easypermissions/EasyPermissions.java

```diff
@@ -85,205 +80,220 @@ public static boolean hasPermissions(@NonNull Context context, @NonNull String..
      *
-     * @param object
-     *         Activity or Fragment requesting permissions. Should implement
-     *         {@link android.support.v4.app.ActivityCompat.OnRequestPermissionsResultCallback}
-     *         or {@code android.support.v13.app.FragmentCompat.OnRequestPermissionsResultCallback}
-     * @param rationale
-     *         a message explaining why the application needs this set of permissions, will be displayed if the user rejects the request the first
-     *         time.
-     * @param positiveButton
-     *         custom text for positive button
-     * @param negativeButton
-     *         custom text for negative button
-     * @param requestCode
-     *         request code to track this request, must be < 256.
-     * @param perms
-     *         a set of permissions to be requested.
+     * @param activity       Activity requesting permissions. Should implement {@link
+     *                       ActivityCompat.OnRequestPermissionsResultCallback} {@code
+     *                       android.support.v13.app.FragmentCompat.OnRequestPermissionsResultCallback}
+     * @param rationale      a message explaining why the application needs this set of permissions,
+     *                       will be displayed if the user rejects the request the first time.
+     * @param positiveButton custom text for positive button
+     * @param negativeButton custom text for negative button
+     * @param requestCode    request code to track this request, must be < 256.
+     * @param perms          a set of permissions to be requested.
+     * @see android.Manifest.permission
      */
-    @SuppressLint("NewApi")
-    public static void requestPermissions(@NonNull final Object object, @NonNull String rationale,
-                                          @StringRes int positiveButton, @StringRes int negativeButton,
-                                          final int requestCode, @NonNull final String... perms) {
-
-        checkCallingObjectSuitability(object);
-
-        // Determine if rationale should be shown (generally when the user has previously
-        // denied the request);
-        boolean shouldShowRationale = false;
-        for (String perm : perms) {
-            shouldShowRationale =
-                    shouldShowRationale || shouldShowRequestPermissionRationale(object, perm);
+    @SuppressWarnings("NewApi")
+    public static void requestPermissions(@NonNull Activity activity,
+                                          @NonNull String rationale,
+                                          @StringRes int positiveButton,
+                                          @StringRes int negativeButton,
+                                          int requestCode,
+                                          @NonNull String... perms) {
+        if (hasPermissions(activity, perms)) {
+            notifyAlreadyHasPermissions(activity, requestCode, perms);
+            return;
         }
```

Comment:
@samtstern I got confused by this too at first, but think of it this way: anytime you call `requestPermissions`, you can expect to get a callback. In your example the flow is still the same: `no permissions -> request permissions -> foo gets called because user accepted permissions -> has permissions -> don't call request permissions -> continue as usual`. The only case I can think of that results in the method being recursively called infinity is if you do this which is wrong anyway:
```java
@AfterPermissionsGranted(foo)
public void bar() {
    EasyPermissions.requestPermissions(this, ...);
    
    // Boom!
}
```

The usage I can see for getting a callback when you already have the permissions would be:
```java
@Override
public void onCreate(instance) {
    EasyPermissions.requestPermissions(this, ...);
}

@AfterPermissionsGranted(foo)
public void bar() {
    // Happily do stuff
}
```